### ISSUE TITLE
unexpected auto-casting / use input references

### ISSUE THREAD

**Issue by glample** [NVIDIA/apex#385]
Hi,

I'm facing two strange issues when I use Apex with fp16:

1) When I use an encoder-decoder architecture, at the end of my encoder forward function the type of my tensor is fp16, as expected. But if I do `h = encoder(x)` then `h` is of type fp32, although the type was fp16 in the forward function just before I return `h`. This is a bit weird and then I have to do something like: `output = decoder(h.half())`. Is that expected?

2) Sometimes I provide as input a tensor and a dictionary to the encoder: `h = encoder(x, dico)` and the forward function of my encoder modifies the dictionary. But everything happens like if I was doing `h = encoder(x, copy.deepcopy(dico))`, i.e. the dictionary is identical when I exit the forward function.

Any ideas?
Thank you

**Comment by ptrblck** [comment]
Hi @glample,

1. You shouldn't cast the input manually to `.half()`, as `amp` will do it automatically for you. The `decoder` should therefore also use FP16 tensors in its forward (same as `encoder`).

2. Could you explain the use case a bit more? What are you passing in the `dict`? Are you passing some kind of data which should be manipulated inplace or some arguments?

**Comment by glample** [comment]
Hi, thanks for your reply.

1) So the problem is that if I don't manually call `h.half()` then I have that `h = encoder(x)` is a fp32 tensor, so I cannot give it to the decoder directly because if I do, in the decoder forward function PyTorch raises an error that tells me that the input is fp32 and the model expects fp16 (all encoder and decoder weights are in fp16). `amp` does everything in fp16 in the encoder, as expected, but somehow converts the output back to fp32 and I can't figure out why.

2) The use case is for instance when I do generation with a Transformer language model (I trained my model without issues, and at test time I want to generate sentences). I typically use a `cache` dictionary that contains the output of the decoder at the previous timesteps. So the data in the dictionary is populated in the forward function of the decoder, and is static (it's not manipulated in any way, and I'm at test time so all is in `no_grad` mode).
More precisely, this is in this code: https://github.com/facebookresearch/XLM/blob/master/src/model/transformer.py#L459

The dictionary is initialized to `cache = {'slen': 0}` and remains unchanged, although I update it in the decoder. It works in fp32 without `amp` , and also in fp16 with the previous apex API (I want to update the codebase with `amp`, but currently I have these two issues).

**Comment by mcarilli** [comment]
By default, amp patches the `.forward` method of models passed through `amp.initialize` to cast any inputs and outputs to fp32, including recursively casting inputs/outputs that are Python containers (ie dicts). The idea behind this was to ensure that all user-visible portions of the code run in fp32. 

The output type casting can be changed using the [cast_model_outputs](https://nvidia.github.io/apex/amp.html#module-apex.amp) argument to `amp.initialize`.  Specifically, if you are passing `encoder` directly to `amp.initialize`, you can try supplying `cast_model_types=torch.half`.  If you do so, you may also wish to insert a manual float cast once your "final" output (the output of the decoder) is produced (before you compute your loss function) because the loss function computation should be done in fp32.

The dict is getting deepcopied here:  https://github.com/NVIDIA/apex/blob/master/apex/amp/_initialize.py#L45-L46.  To avoid the deepcopy, you may need to maintain the dict at whatever scope surrounds the encoder/decoder, pass it by assigning as an attribute rather than passing it as an argument, or change [line 134](https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L134) to just `return value`.  Maybe the best solution is for us to change the casting logic to update any passed dicts in-place rather than via a deepcopy (I wrote it that way because I was following following the pattern in [torch.optim.optimizer](https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L133-L134)).

FYI checkpointing got put on the back burner due to more internal optimization work and also work towards putting Amp upstream.  I think ultimately Amp needs to be upstream, although Apex will continue to be maintained.

**Comment by glample** [comment]
Thanks! That makes sense. I was able to fix the dictionary problem with an attribute assignment as you suggested.

Did you mean to set `cast_model_outputs=torch.half` instead of `cast_model_types=torch.half`?
I just tried this and it also solved the problem, but it somehow made the code slower than manually converting to `half` the encoder output, I'm not sure why (maybe the casting involves a lot of others conversions that I cannot see), so I will stick to the original solution.

Btw, why doing this automatic casting of inputs/outputs? Will it not lead to a lot of half->float32->half castings that may slow down the overall network? Wouldn't it be enough to patch the inputs/outputs of functions like softmax, cross_entropy, etc?
